BUG: pd.show_versions: json.decoder.JSONDecodeError

alexprincel opened the pull request:
- [ x] closes #39701
- [ x] tests added / passed
- [ x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

alexprincel committed:
BUG: Fix pd.show_versions as_json invalid JSON (#39701)

simonjayhawkins commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,21 +33,44 @@
 )
 @pytest.mark.filterwarnings("ignore:Setuptools is replacing distutils:UserWarning")
-def test_show_versions(capsys):
+@pytest.mark.parametrize("as_json", [True, False, "test_output.json"], indirect=True)
```
why do we need the indirect parameterisation of the fixture? There are no composite fixtures in play here?

probably better to parametrize the fixture directly, or just parameterise the test function and remove the fixture.

jreback replied:
yeah this is only used once, just paramterize directly

simonjayhawkins commented on pandas/util/_print_versions.py:
```diff
@@ -107,7 +107,7 @@ def show_versions(as_json: Union[str, bool] = False) -> None:
 
         if as_json is True:
-            print(j)
+            print(json.dumps(j))
```
when printing to file, `json.dump(j, f, indent=2)` is used. would it be better to have the console output and file output the same?

maybe change the test, to ensure this consistency.

simonjayhawkins reviewed (commented):
Thanks @alexprincel for the PR.

lithomas1 reviewed (commented):
Can you add a note in the whatsnew?(```doc/source/whatsnew/v1.3.0.rst```, I think this should go under Other in the section bug fixes.)

jreback commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,21 +33,44 @@
-    assert "INSTALLED VERSIONS" in result
+    # check valid json is printed to the console if as_json is True
+    if as_json is True:
+        json.loads(result)
```
can you assert somethings here

jreback reviewed (changes requested):
pls also add a whatsnew note. bug fixes in I/O sectoin are ok

pep8speaks commented:
Hello @alexprincel! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:













* In the file [`pandas/tests/util/test_show_versions.py`](https://github.com/pandas-dev/pandas/blob/3dfc16ee501424b4c6822693f3364a8d7a1d9678/pandas/tests/util/test_show_versions.py):

> [Line 78:39](https://github.com/pandas-dev/pandas/blob/3dfc16ee501424b4c6822693f3364a8d7a1d9678/pandas/tests/util/test_show_versions.py#L78): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 78:41](https://github.com/pandas-dev/pandas/blob/3dfc16ee501424b4c6822693f3364a8d7a1d9678/pandas/tests/util/test_show_versions.py#L78): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 78:71](https://github.com/pandas-dev/pandas/blob/3dfc16ee501424b4c6822693f3364a8d7a1d9678/pandas/tests/util/test_show_versions.py#L78): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 78:73](https://github.com/pandas-dev/pandas/blob/3dfc16ee501424b4c6822693f3364a8d7a1d9678/pandas/tests/util/test_show_versions.py#L78): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 78:89](https://github.com/pandas-dev/pandas/blob/3dfc16ee501424b4c6822693f3364a8d7a1d9678/pandas/tests/util/test_show_versions.py#L78): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 88 characters)
> [Line 80:46](https://github.com/pandas-dev/pandas/blob/3dfc16ee501424b4c6822693f3364a8d7a1d9678/pandas/tests/util/test_show_versions.py#L80): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

alexprincel commented:
Not sure if this needs commenting (again, first-timer), but I believe I have implemented all requested changes and made testing more robust.

lithomas1 commented:
@alexprincel Looking at the diff, it looks like a merge went wrong somewhere. Can you fix this up?

alexprincel commented:
> @alexprincel Looking at the diff, it looks like a merge went wrong somewhere. Can you fix this up?

Sorry about that. What seems to be the problem? I will do my best to resolve it. My git skills are not yet up to par.

lithomas1 commented:
@alexprincel If you click on ```Files changed```, you can see that there are unrelated changes in the ```pandas/tests/io``` directory. (Which I suspect is the result of a bad merge) I'm not really an expert on git either, so maybe you can ask one of the other maintainers for help.(If necessary you can revert the changes manually)

alexprincel commented:
> @alexprincel If you click on `Files changed`, you can see that there are unrelated changes in the `pandas/tests/io` directory. (Which I suspect is the result of a bad merge) I'm not really an expert on git either, so maybe you can ask one of the other maintainers for help.(If necessary you can revert the changes manually). Apart from the unrelated changes, LGTM.

I saw this as well after. I have been reading on git this week and trying some things on test repos just to understand exactly what I need to do to correct this. If I can't figure it out I'll just start a new branch, put the modified files in that branch's working directory and commit that.

MarcoGorelli commented:
If you only need to revert a handful of files, I'd suggest
```bash
git remote -v  # check that `upstream` is pandas-dev/pandas and `origin` is alexprincel/pandas
git fetch upstream master
git checkout upstream/master -- pandas/tests/io
git commit -m 'revert unrelated changes'
git push origin alexdev
```

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,21 +30,65 @@
+def test_show_versions(capsys, as_json, tmpdir):
     # gh-32041
-    pd.show_versions()
+    if isinstance(as_json, str):
+        as_json = os.path.join(tmpdir, as_json)
```
This test is hard to follow can you make separate tests? See [don't put logic in tests](https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html)

alexprincel replied:
Thank you for those comments! Will implement them as soon as I figure out why isort is failing here but not on my local environment..

MarcoGorelli replied:
my guess is that you haven't fetched and merged latest upstream changes, there have been some recent changes

alexprincel replied:
I have separated it into multiple smaller functions in my local branch now. I'm looking at all the warning silencing that is done and wondering whether this is still useful. Looking at the git blame, the original author added this saying it would remove some warnings during CI. I tested removing those warning silencing and running pre-commit on my end and I do not see any warnings. Similarly, other test files do not have this warning. Should I go ahead and remove the warning filtering ?

MarcoGorelli replied:
@alexprincel which warnings are you referring to? pre-commit just runs linting/formatting jobs, it wouldn't pick up on warnings given by pandas

alexprincel replied:
I'm referring to the multiple calls to decorator @pytest.mark.filterwarnings. I am still unfamiliar with some aspects of purest, amongst other things, but removing those decorators does not result in warnings being raised when running pytest on test_show_version.

MarcoGorelli replied:
Ah OK, previously you'd written

> I tested removing those warning silencing and running pre-commit on my end

I guess you meant `pytest` - now this makes sense to me, thanks

alexprincel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into alexdev

alexprincel committed:
BUG: Separate multiple tests (39701)

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -44,3 +94,15 @@ def test_show_versions(capsys):
+    out_path = os.path.join(tmpdir, "test_json.json")
+    pd.show_versions(as_json=out_path)
+    with open(out_path) as out_fd:
+        result_file = "".join(out_fd.readlines())
```
is this the same as `out_fd.read()`? If so, then maybe that's a bit simpler?

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,13 +33,56 @@
+    # make sure that the file was created
+    assert os.path.exists(as_json)
+
+    with open(as_json) as fd:
+        contents = fd.readlines()
+        str_contents = "".join(contents)
+
+        # make sure that there was output to the file
+        assert str_contents
+
+        # check if file output is valid JSON, will raise an exception if not
+        dict_check = json.loads(str_contents)
```
any reason to do this over
```python
    with open(as_json) as fd:
        dict_check = json.load(fd)
```
?

Also, I think the naming is a bit confusing, maybe rename `dict_check` to `result` and `version_elements` to `expected`

alexprincel replied:
None that I can think of. It seems I made this more complicated than it need to be. Thanks !

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,13 +33,56 @@
+    as_json = os.path.join(tmpdir, "test_output.json")
+
+    pd.show_versions(as_json=as_json)
+
+    # make sure that the file was created
+    assert os.path.exists(as_json)
```
the next lines will fail if this isn't true, so this strikes me as redundant, I think it can be removed

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,13 +33,56 @@
     captured = capsys.readouterr()
     result = captured.out
 
-    # check header
```
why remove this comment?

alexprincel committed:
BUG: Various minor improvements on test_show_versions (#39701)

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,11 +37,44 @@
+
+def test_show_versions_console_json(capsys):
+    pd.show_versions(as_json=True)
+    result = get_captured_output(capsys)
+
+    # check valid json is printed to the console if as_json is True
+    result = json.loads(result)
```
instead of reassigning to `result` (which may cause problems when we expand mypy checks on tests) could you just do something like
```
stdout = capsys.readouterr().out
result = json.loads(stdout)
```
I'm not sure that making a function for `get_captured_output` adds much, repetition is more acceptable in tests than it is in source code

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,11 +37,44 @@
 )
 @pytest.mark.filterwarnings("ignore:Setuptools is replacing distutils:UserWarning")
-def test_show_versions(capsys):
+def test_show_versions(tmpdir):
```
below every test, can you add a comment with the issue number? something like
```
# GH39701
```
, see other recent tests

MarcoGorelli reviewed (changes requested):
Thanks @alexprincel for updating - I think this almost ready, just a couple of minor comments

alexprincel commented:
> Thanks @alexprincel for updating - I think this almost ready, just a couple of minor comments

Will make them later today. Thanks for your comments, helps me improve.

alexprincel committed:
BUG: Various test improvements to test_show_versions (#39701)

alexprincel commented:
I have addressed all comments. However, three errors  keep coming up during CI and I can't figure out why. pandas-dev.pandas (+ Linux py37_np_dev + Linux py39) all fail but I do not understand the error messages at all.

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -1,10 +1,21 @@
 import pandas as pd
 
 
+def get_captured_output(capsys):
+    return capsys.readouterr().out
```
This is unused, can you remove it please?

MarcoGorelli reviewed (changes requested):
Thanks @alexprincel - just one unused function to remove, other than that this looks good to me

alexprincel committed:
BUG: Remove unused function (#39701)

MarcoGorelli commented on pandas/tests/util/test_show_versions.py:
```diff
@@ -26,11 +33,47 @@
-    pd.show_versions()
-    captured = capsys.readouterr()
-    result = captured.out
+    # GH39701
```
I should've picked up on this earlier, but this one didn't need changing, as it was in the existing test
```suggestion
    # gh-32041
```

alexprincel committed:
Update github issue number

Co-authored-by: Marco Gorelli <marcogorelli@protonmail.com>

MarcoGorelli reviewed (approved):
Pending green, this looks good to me

@simonjayhawkins any further comments, or good to merge?

simonjayhawkins commented:
LGTM

MarcoGorelli commented:
cool - merging then, as commit `02e6bc7` was green and the commit following that just changed a comment

Thank you @alexprincel !

alexprincel commented:
> cool - merging then, as commit `02e6bc7` was green and the commit following that just changed a comment
> 
> Thank you @alexprincel !

Thank you for your patience on this! Hopefully I will be more efficient on my next PR.
